### PR TITLE
refactor: update tui-code-snippet@2.2.0

### PR DESCRIPTION
**Pull request by dotaitch**
* set webpack optimization
* apply toast-ui.release-notes
* apply tui-code-snippet@2.2.0
* refactor examples
* edit description of tests
* convert examples in README and getting-started to ES6

### CODE REVIEW

File: src/js/features/ajax.js [hunk 1]
```diff
@@ -126,7 +132,7 @@ var Ajax = snippet.defineClass(
        */
       if (
         !this.tree.invoke('beforeAjaxRequest', {
-          type: type,
+          type: type, // TODO: change to `command`
```
Comment (dotaitch) [inline src/js/features/ajax.js]: Ajax 관련 이벤트 (`beforeAjaxRequest`, `errorAjaxResponse`, `failAjaxResponse`, `successAjaxResponse`)에서 반환하는 데이터가 [API 문서](https://nhn.github.io/tui.tree/3.5.6/Tree#event-beforeAjaxRequest)에는 `command`라고 적혀있으나, 실제는 `type`으로 넘어오고 있습니다. 추후 Ajax 모듈 교체하면서 메이저 업데이트를 할 예정인데, 그 때 `type`을 `command`로 고쳐도 될까요?

File: src/js/features/contextMenu.js [hunk 2]
```diff
@@ -219,7 +218,7 @@ var ContextMenu = snippet.defineClass(
        * });
        */
       this.tree.fire('selectContextMenu', {
-        cmd: cmd,
+        cmd: cmd, // TODO: change to `command`
```
Comment (dotaitch) [inline src/js/features/contextMenu.js]: 타 이벤트는 `command`로 되어있는데, `selectContextMenu`만 `cmd`를 넘기고 있습니다. 다른 이벤트들과 동일하게 바꾸는 것(`cmd` -> `command`)이 좋을 것 같은데 어떻게 생각하시나요?

File: src/js/tree.js [hunk 3]
```diff
@@ -26,12 +44,11 @@ var features = {
   ContextMenu: ContextMenu,
   Ajax: Ajax
 };
-var snippet = require('tui-code-snippet');
-var extend = snippet.extend;
 
 var TIMEOUT_TO_DIFFERENTIATE_CLICK_AND_DBLCLICK = 200;
 var MOUSE_MOVING_THRESHOLD = 5;
 
+// TODO: specify tui-code-snippet's version (package.json)
```
Comment (dotaitch) [inline src/js/tree.js]: 추후 tui-code-snippet@2.3.0 적용 및 Ajax 모듈을 교체하면서 하면서 버전을 명시할 예정입니다

File: src/js/features/draggable.js [hunk 4]
```diff
@@ -370,16 +353,16 @@ var Draggable = snippet.defineClass(
...
+      var target = getTarget(event);
       var nodeId;
 
       if (!this.useHelper) {
         return;
       }
 
       this._setClassNameOnDragItem('add');
-      this._changeHelperPosition(mousePos);
+      this._changeHelperPosition(mousePos); // TODO: test Failed
```
Comment (dotaitch) [inline src/js/features/draggable.js]: TODO 주석 제거할 예정입니다

File: examples/example08-ajax.html [hunk 5]
```diff
@@ -156,31 +154,35 @@
...
-            var message;
+        tree.on('beforeMove', function() {
+            console.log('before move');
+        });
 
-            message = 'Command : ' + command  + '\n';
-            message += 'jQuery request options : \n';
-            message += JSON.stringify(tree.enabledFeatures.Ajax.command[command], null, 8);
+        tree.on('successAjaxResponse', function(eventData) {
+            var command = eventData.type; // TODO: change to `command`
```
Comment (seonim-ryu) [inline examples/example08-ajax.html]: `command`가 파라미터로 넘어오지 않나요?? API가 변경된건가요 아니면 예제가 잘못되어 있었던건가요?

File: src/js/features/ajax.js (hunk 1 above)
Comment (seonim-ryu) [reply src/js/features/ajax.js]: 예제 페이지에서 질문했던 내용이 여기 있었군요ㅎ
일단 현재 doc에 노출되어 있는 정보는 수정이 필요하고요. `command` -> `type`
Ajax 모듈 교체할 때는 네이밍 변경이 필요한데, `command`가 될지 아니면 다른 네이밍을 사용할지는 고려해봐야겠네요.

Comment (seonim-ryu) [reply src/js/features/ajax.js]: 아니면 지금 `command`로 변경하고 `type`을 디프리케이트 하는 방법도 있겠네요ㅎ
`commamd`을 사용할지는 검토 한 번 부탁드릴게요~
```js
{
  type: type, // @deprecated
  command: type,
  data: data
}
```

File: src/js/features/contextMenu.js (hunk 2 above)
Comment (seonim-ryu) [reply src/js/features/contextMenu.js]: 이것도 위에 이벤트 객체의 `type` 프로퍼티와 같은 이슈네요.
기존에 `cmd`가 컨텍스트에서 넘어온 객체의 프로퍼티명을 그대로 사용해서 통일이 안된듯해요.
네이밍 변경 찬성합니다. 단, 이번에 디프리케이트 처리할 예정이라면 네이밍 한 번 검토해주세요~

File: src/js/features/draggable.js [hunk 6]
```diff
@@ -205,8 +199,8 @@ var Draggable = snippet.defineClass(
       var helperStyle = this.helperElement.style;
       var pos = this.tree.rootElement.getBoundingClientRect();
 
-      helperStyle.top = mousePos.y - pos.top + this.helperPos.y + 'px';
-      helperStyle.left = mousePos.x - pos.left + this.helperPos.x + 'px';
+      helperStyle.top = mousePos[1] - pos.top + this.helperPos.y + 'px';
+      helperStyle.left = mousePos[0] - pos.left + this.helperPos.x + 'px';
```
Comment (seonim-ryu) [inline src/js/features/draggable.js]: 여기 반환되는 값이 변경되었나요??
> https://github.com/nhn/tui.tree/blob/d8e2c91e0eb3611dbed9faa022af82d99524b36e/src/js/features/draggable.js#L10

File: src/js/features/draggable.js [hunk 7]
```diff
@@ -23,15 +38,6 @@ var defaultOptions = {
   isSortable: false
 };
 var rejectedTagNames = ['INPUT', 'BUTTON', 'UL'];
-var selectKey = util.testProp([
-  'userSelect',
-  'WebkitUserSelect',
-  'OUserSelect',
-  'MozUserSelect',
-  'msUserSelect'
-]);
```
Comment (seonim-ryu) [inline src/js/features/draggable.js]: 기본 셀렉션을 막을 때 사용하는 값인데, 제거된 이유가 따로 있나요?

File: src/js/treeModel.js [hunk 8]
```diff
@@ -155,15 +158,24 @@ var TreeModel = snippet.defineClass(
...
     getCount: function() {
-      return keys(this.treeHash).length;
+      var treeHash = this.treeHash;
+      var length = 0;
+      var node;
+
+      // eslint-disable-next-line guard-for-in
+      for (node in treeHash) {
+        length += treeHash.hasOwnProperty(node);
+      }
```
Comment (seonim-ryu) [inline src/js/treeModel.js]: 린트 룰 disable만으로 처리하면 안될 것 같습니다.
`keys` 라는 함수를 유틸 모듈에 하나 추가하고 객체에 프로퍼티를 가지고 있는지 체크하도록 변경하는게 맞지 않을까요??

* 참고 : https://eslint.org/docs/rules/guard-for-in
```js
for (key in foo) {
    if (Object.prototype.hasOwnProperty.call(foo, key)) {
        doSomething(key);
    }
}
```

File: test/features/contextMenu.spec.js [hunk 9]
```diff
@@ -1,11 +1,8 @@
 var Tree = require('../../src/js/tree');
-var util = require('../../src/js//util');
 var TuiContextMenu = require('tui-context-menu');
-var styleKeys = ['userSelect', 'WebkitUserSelect', 'OUserSelect', 'MozUserSelect', 'msUserSelect'];
-var enableProp = util.testProp(styleKeys);
 
 describe('contextMenu.js', function() {
```
Comment (seonim-ryu) [inline test/features/contextMenu.spec.js]: `contextMenu.js` -> `context menu feature`로 변경하는게 좋겠네요ㅎ (왜 저렇게 했었을까...)

File: test/index.js [hunk 10]
```diff
@@ -1,5 +1,5 @@
 var testsContext = require.context('.', true, /spec\.js$/);
-var featureTestsContext = require.context('./features', true, /spec\.js$/);
+var featureTestsContext = require.context('./features', true, /specx\.js$/);
```
Comment (seonim-ryu) [inline test/index.js]: 이건.. 오타인건가요??

File: bower.json [hunk 11]
```diff
@@ -15,14 +15,15 @@
...
+        "tuidoc.config.json",
+        "webpack.config.js"
     ],
     "dependencies": {
-        "tui-code-snippet": "^1.5.0",
-        "tui-context-menu": "^2.1.1"
+        "tui-context-menu": "^2.1.6"
     },
     "devDependencies": {
         "jquery": "^1.11.0"
```
Comment (seonim-ryu) [inline bower.json]: 음.. jQuery도 `dependencies`로 올라가는게 맞겠어요.
`tui-context-menu`, `jquery` 둘 다 옵셔널로 사용되는 것들인데 컨텍스트만 디펜던시로 올라가 있는게 이상하네요. ;ㅅ;

File: docs/getting-started.md [hunk 12]
```diff
@@ -81,7 +75,7 @@ var options = {
 var tree = new tui.Tree('tree', options);
```
Comment (seonim-ryu) [inline docs/getting-started.md]: 여기도 예제를 모듈 import 하는 방식으로 변경하면 좋겠습니다~

File: src/js/features/draggable.js (hunk 6 above)
Comment (dotaitch) [reply src/js/features/draggable.js]: 원래 코드는 tree 컴포넌트 안에 정의되어있던 [util.js의 getMousePos](https://github.com/nhn/tui.tree/blob/v3.5.6/src/js/util.js#L322)를 사용 중이었습니다. 이 util이 [tui-code-snippet의 getMousePosition](https://nhn.github.io/tui.code-snippet/2.2.0/domEvent#getMousePosition)과 기능이 동일하다고 생각되어 tui-code-snippet으로 교체하였습니다.
그런데 util.js가 **x, y 프로퍼티를 가진 객체**를 반환하는 것과 달리 tui-code-snippet은 **x가 첫번째, y가 두번째 요소인 배열**를 반환하여서 이렇게 변경하였습니다.

File: src/js/features/draggable.js (hunk 7 above)
Comment (dotaitch) [reply src/js/features/draggable.js]: tui-code-snippet의 `disableTextSelection`과 `enableTextSelection`으로 교체하였고, testProp은 이 두가지 함수 내에 포함되어 있어서 제거하였습니다.
* https://github.com/nhn/tui.code-snippet/blob/v2.2.0/domUtil/disableTextSelection.js#L15
* https://github.com/nhn/tui.code-snippet/blob/v2.2.0/domUtil/enableTextSelection.js#L16

File: test/index.js (hunk 10 above)
Comment (dotaitch) [reply test/index.js]: 앗 네 오타예요 Test가 모두 통과하는 건 확인하였고, 코드 리뷰 적용하면서 다시 수정해놓겠습니다.

File: src/js/features/draggable.js (hunk 6 above)
Comment (jung-han) [reply src/js/features/draggable.js]:
```js
var mousePosX = mousePos[0];
var mousePosY = mousePos[1];

helperStyle.top = mousePosY - pos.top + this.helperPos.y + 'px';
helperStyle.left = mousePosX - pos.left + this.helperPos.x + 'px';
```

이렇게 작성하는건 어떤가요?_?

File: src/js/tree.js [hunk 13]
```diff
@@ -390,22 +409,24 @@ var Tree = snippet.defineClass(
...
-     * @param {MouseEvent} event - Click event
+     * @param {MouseEvent} ev - Click event
      * @private
      */
-    _onClick: function(event) {
-      var target = util.getTarget(event);
-      var self = this;
+    _onClick: function(ev) {
+      var target = getTarget(ev);
+      var isRightButton = getMouseButton(ev) === 2;
```
Comment (jung-han) [inline src/js/tree.js]: 2는 상수로 만들어서 관리하는게 좋을 것 같습니다

File: test/tree.spec.js [hunk 14]
```diff
@@ -66,17 +67,17 @@ describe('Tree', function() {
...
       var textNode = util.getFirstTextNode(textElement);
 
       expect(textNode.nodeValue).toEqual('A');
     });
 
     it('should make DOM from optional-template', function() {
-      var $leafNodes = $rootElement.find('.tui-tree-leaf'),
-        $leafLabels = $leafNodes.find('.tui-tree-leaf-label');
+      var leafNodes = rootElement.querySelectorAll('.tui-tree-leaf'),
+        leafLabels = rootElement.querySelectorAll('.tui-tree-leaf .tui-tree-leaf-label');
```
Comment (jung-han) [inline test/tree.spec.js]: 변수를 구분해서 작성해 주셔야 할 것 같습니다.
`.tui-tree-leaf`를 추가해준 이유가 있을까요?_?

Comment (dotaitch) [reply test/tree.spec.js]: `leafLabels`는 모든 leafNode 하위에 있는 라벨들 입니다. leafNodes가 하나 이상이기 때문에 만약 `.tui-tree-leaf-label`만으로 검색하려면 루프를 이용하여 leafNodes 하나씩 탐색해야합니다. 루프를 돌며 탐색하는 과정을 없애기 위해 이렇게 바꾸었는데, 루프를 이용하는 것이 더 좋은 방법인가요?

File: src/js/features/selectable.js [hunk 15]
```diff
@@ -83,14 +90,14 @@ var Selectable = snippet.defineClass(
...
-      var target = util.getTarget(event),
+      var target = getTarget(event),
         nodeId = this.tree.getNodeIdFromElement(target);
 
       this.select(nodeId, target);
     },
 
     /* eslint-disable valid-jsdoc */
-    /* Ignore "target" parameter annotation for API page
+    /* Ignore "target" parameter annotation for API page // TODO: question
```
Comment (dotaitch) [inline src/js/features/selectable.js]: 여기서 왜 `target`을 API 문서에 표기하지 않나요? 표기하지 않는다면 사용자는 `target`을 넘기지 않을 것이고, 그럼 `beforeSelect`과 `select` 이벤트의 `target`은 항상 `undefined` 일텐데 이렇게 한 이유가 있나요?
그리고 `target`은 `select` 함수 내에서 별다른 쓰임이 없고 `beforeSelect`과 `select` 이벤트 발생 시 넘어오는 상태 그대로 넘겨주는 용도로만 사용되는데, 차라리 없애는 게 좋지 않을까요?

File: README.md [hunk 16]
```diff
@@ -148,23 +149,23 @@ To get the constructor function, you should import the module using one of the f
...
 
 #### Using module format in node environment
 ``` javascript
-var Tree = require('tui-tree'); /* CommonJS */
+const Tree = require('tui-tree'); /* CommonJS */
 ```
 
 ``` javascript
-import {Tree} from 'tui-tree'; /* ES6 */
+import Tree from 'tui-tree'; /* ES6 */
```
Comment (seonim-ryu) [inline README.md]: 👍

File: src/js/features/ajax.js (hunk 1 above)
Comment (dotaitch) [reply src/js/features/ajax.js]: Ajax 이벤트에서 반환하는 데이터는 [enableFeature의 Ajax](https://nhn.github.io/tui.tree/latest/Tree#enableFeature)를 정의할 때 쓰이는 `command`의 키 값(ex. read, create, update, remove) 입니다. 이 옵션과 동일한 이름인 `command`를 사용하는게 좋을 것 같은데, 어떻게 생각하시나요?

File: src/js/features/selectable.js (hunk 15 above)
Comment (dotaitch) [reply src/js/features/selectable.js]: `select` API를 private과 public으로 나누었습니다.
